### DISCUSSION TITLE
How to Log Compiled SQL Statements with Parameters Inlined Using SQLAlchemy with mssql+aioodbc

### DISCUSSION THREAD

**Discussion by ben-sandham** [sqlalchemy/sqlalchemy#11962] [Usage Questions] [answered]
I'm currently working on a project where I'm using SQLAlchemy with the mssql+aioodbc dialect to connect to a local Azure SQL Edge database. For debugging purposes, I need to have my engine connection output the compiled SQL statements with the actual parameter values filled in—that is, with literal binds instead of placeholders like `?`.

I've tried several approaches, including setting echo='debug' and adjusting the logging configurations, but the SQL statements in the logs still contain `?` placeholders, and the parameters are logged separately. 

1. Is there a way to configure SQLAlchemy to echo the compiled SQL statements with parameters binded when using the mssql+aioodbc dialect?
2. If not directly possible, are there any recommended workarounds or tools that can help achieve this for debugging purposes?
3. Is this limitation inherent to the MSSQL dialect with ODBC drivers, and if so, are there plans to support this feature in future releases?

**Comment by CaselIT** [comment, marked as answer]
Hi,

1. no
2. you need to manually compile the statements. See https://docs.sqlalchemy.org/en/20/faq/sqlexpressions.html#how-do-i-render-sql-expressions-as-strings-possibly-with-bound-parameters-inlined for details
3. it's not something that's supported by sqlalchemy. Main reason is that not all types have literal bind support.
